server_development/topics/user-storage/stream-interfaces.adoc

```diff
@@ -0,0 +1,48 @@
+
+=== Stream-based Interfaces
+
+Many of the user storage interfaces in {project_name} contain query methods that can return potentially large sets of objects,
+which might lead to in significant impacts in terms of memory consumption and processing time - specially when only a small subset of the
+objects' internal state is used in the query method's logic.
+
+In order to provide developers with a more efficient alternative to process large data sets in these query methods, a `Streams`
+sub-interface has been added to user storage interfaces. These `Streams` sub-interfaces replace the original collection-based
+methods in the super-interfaces with stream-based variants, making the collection-based methods default (they invoke their `Stream`
+counterpart and collect the result).
+
+The `Streams` sub-interfaces allows for implementations to focus on the stream-based approach for processing sets of data and
+benefit from the potential memory and performance optimizations of that approach. The interfaces that offer a `Streams`
+sub-interface to be implemented include a few _capability interfaces_, all interfaces in the `org.keycloak.storage.federated`
+package and a few others that might be implemented depending on the scope of the custom storage implementation.
+
+You can find bellow a list of the interfaces that offer a `Streams` sub-interface to developers.
+
+|===
+| Package | Classes
+| `org.keycloak.credential` | `CredentialInputUpdater`(*), `UserCredentialStore`
+| `org.keycloak.models` | `GroupModel`, `RoleMapperModel`, `UserCredentialManager`, `UserModel`, `UserProvider`
+| `org.keycloak.models.cache` | `CachedUserModel`, `UserCache`
+| `org.keycloak.storage.federated` | All classes
+| `org.kecyloak.storage.user` | `UserQueryProvider`(*)
+|===
+(*) identify _capability interfaces_
+
+To make use of the `Streams` variants, one has to simply implement these sub-interfaces instead of the original interfaces
+when creating a custom provider. For example, the following code uses the `Streams` variant of the `UserQueryProvider` interface:
+
+[source,java]
+----
+public class CustomQueryProvider extends UserQueryProvider.Streams {
+...
+    @Override
+    Stream<UserModel> getUsersStream(RealmModel realm, int firstResult, int maxResults) {
```

Comment:
```suggestion
    Stream<UserModel> getUsersStream(RealmModel realm, Integer firstResult, Integer maxResults) {
```

Reply:
Ah, yes.. I was using my last worked-on branch which doesn't have these changes.. I'm updating the example